# Data Understanding

Initial exploration of the BoardGameGeek dataset.

Goals:
- Understand the available tables
- Inspect shape, columns, and keys
- Identify the main unit of analysis
- Decide how to structure a first modelling dataset

### Imports

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

### Getting the data

In [2]:
DATA_DIR = Path("../raw_data")
DATA_DIR

PosixPath('../raw_data')

In [4]:
for file in sorted(DATA_DIR.iterdir()):
    print(file.name)

.DS_Store
.gitkeep
artists_reduced.csv
bgg_data_documentation.txt
designers_reduced.csv
games.csv
mechanics.csv
publishers_reduced.csv
ratings_distribution.csv
subcategories.csv
themes.csv
user_ratings.csv


In [13]:
games = pd.read_csv(DATA_DIR / "games.csv")
mechanics = pd.read_csv(DATA_DIR / "mechanics.csv")
themes = pd.read_csv(DATA_DIR / "themes.csv")
user_ratings = pd.read_csv(DATA_DIR / "user_ratings.csv")
publishers = pd.read_csv(DATA_DIR / "publishers_reduced.csv")
ratings_distribution = pd.read_csv(DATA_DIR / "ratings_distribution.csv")
subcategories = pd.read_csv(DATA_DIR / "subcategories.csv")

## What does the data look like?

In [6]:
def inspect_df(df, name):
    print(f"--- {name} ---")
    print(f"shape: {df.shape}")
    print("\ncolumns:")
    print(df.columns.tolist())
    print("\nhead:")
    display(df.head())
    print("\nmissing values:")
    display(df.isna().sum().sort_values(ascending=False).head(10))

In [8]:
inspect_df(games, "games")

--- games ---
shape: (21925, 48)

columns:
['BGGId', 'Name', 'Description', 'YearPublished', 'GameWeight', 'AvgRating', 'BayesAvgRating', 'StdDev', 'MinPlayers', 'MaxPlayers', 'ComAgeRec', 'LanguageEase', 'BestPlayers', 'GoodPlayers', 'NumOwned', 'NumWant', 'NumWish', 'NumWeightVotes', 'MfgPlaytime', 'ComMinPlaytime', 'ComMaxPlaytime', 'MfgAgeRec', 'NumUserRatings', 'NumComments', 'NumAlternates', 'NumExpansions', 'NumImplementations', 'IsReimplementation', 'Family', 'Kickstarted', 'ImagePath', 'Rank:boardgame', 'Rank:strategygames', 'Rank:abstracts', 'Rank:familygames', 'Rank:thematic', 'Rank:cgs', 'Rank:wargames', 'Rank:partygames', 'Rank:childrensgames', 'Cat:Thematic', 'Cat:Strategy', 'Cat:War', 'Cat:Family', 'Cat:CGS', 'Cat:Abstract', 'Cat:Party', 'Cat:Childrens']

head:


,BGGId,Name,Description,YearPublished,GameWeight,AvgRating,BayesAvgRating,StdDev,MinPlayers,MaxPlayers,...,Rank:partygames,Rank:childrensgames,Cat:Thematic,Cat:Strategy,Cat:War,Cat:Family,Cat:CGS,Cat:Abstract,Cat:Party,Cat:Childrens
0,1,Die Macher,die macher game seven sequential political rac...,1986,4.3206,7.61428,7.10363,1.57979,3,5,...,21926,21926,0,1,0,0,0,0,0,0
1,2,Dragonmaster,dragonmaster tricktaking card game base old ga...,1981,1.9630,6.64537,5.78447,1.45440,3,4,...,21926,21926,0,1,0,0,0,0,0,0
2,3,Samurai,samurai set medieval japan player compete gain...,1998,2.4859,7.45601,7.23994,1.18227,2,4,...,21926,21926,0,1,0,0,0,0,0,0
3,4,Tal der Könige,triangular box luxurious large block tal der k...,1992,2.6667,6.60006,5.67954,1.23129,2,4,...,21926,21926,0,0,0,0,0,0,0,0
4,5,Acquire,acquire player strategically invest business t...,1964,2.5031,7.33861,7.14189,1.33583,2,6,...,21926,21926,0,1,0,0,0,0,0,0



missing values:


Family                15262
LanguageEase           5891
ComAgeRec              5530
ImagePath                17
Description               1
BGGId                     0
Rank:wargames             0
Kickstarted               0
Rank:boardgame            0
Rank:strategygames        0
dtype: int64

In [9]:
inspect_df(mechanics, "mechanics")

--- mechanics ---
shape: (21925, 158)

columns:
['BGGId', 'Alliances', 'Area Majority / Influence', 'Auction/Bidding', 'Dice Rolling', 'Hand Management', 'Simultaneous Action Selection', 'Trick-taking', 'Hexagon Grid', 'Once-Per-Game Abilities', 'Set Collection', 'Tile Placement', 'Action Points', 'Investment', 'Market', 'Square Grid', 'Stock Holding', 'Victory Points as a Resource', 'Enclosure', 'Pattern Building', 'Pattern Recognition', 'Modular Board', 'Network and Route Building', 'Point to Point Movement', 'Melding and Splaying', 'Negotiation', 'Trading', 'Push Your Luck', 'Income', 'Race', 'Random Production', 'Variable Set-up', 'Roll / Spin and Move', 'Variable Player Powers', 'Action Queue', 'Bias', 'Grid Movement', 'Lose a Turn', 'Programmed Movement', 'Scenario / Mission / Campaign Game', 'Voting', 'Events', 'Paper-and-Pencil', 'Player Elimination', 'Role Playing', 'Movement Points', 'Simulation', 'Variable Phase Order', 'Area Movement', 'Commodity Speculation', 'Cooperative 

,BGGId,Alliances,Area Majority / Influence,Auction/Bidding,Dice Rolling,Hand Management,Simultaneous Action Selection,Trick-taking,Hexagon Grid,Once-Per-Game Abilities,...,Contracts,Passed Action Token,King of the Hill,Action Retrieval,Force Commitment,Rondel,Automatic Resource Growth,Legacy Game,Dexterity,Physical
0,1,1,1,1,1,1,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,2,0,0,0,0,0,0,1,0,0,...,0,0,0,0,0,0,0,0,0,0
2,3,0,1,0,0,1,0,0,1,1,...,0,0,0,0,0,0,0,0,0,0
3,4,0,1,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,5,0,0,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0



missing values:


BGGId                 0
TableauBuilding       0
Mancala               0
Ownership             0
Kill Steal            0
Hidden Movement       0
Track Movement        0
Deck Construction     0
Drafting              0
Prisoner's Dilemma    0
dtype: int64

In [10]:
inspect_df(themes, "themes")

--- themes ---
shape: (21925, 218)

columns:
['BGGId', 'Adventure', 'Fantasy', 'Fighting', 'Environmental', 'Medical', 'Economic', 'Industry / Manufacturing', 'Transportation', 'Science Fiction', 'Space Exploration', 'Civilization', 'Civil War', 'Movies / TV / Radio theme', 'Novel-based', 'Age of Reason', 'Mythology', 'Renaissance', 'American West', 'Animals', 'Modern Warfare', 'Medieval', 'Ancient', 'Nautical', 'Post-Napoleonic', 'Horror', 'Farming', 'Religious', 'Travel', 'Murder/Mystery', 'Pirates', 'Comic Book / Strip', 'Mature / Adult', 'Video Game Theme', 'Spies/Secret Agents', 'Arabian', 'Prehistoric', 'Trains', 'Aviation / Flight', 'Zombies', 'World War II', 'Racing', 'Pike and Shot', 'World War I', 'Humor', 'Sports', 'Mafia', 'American Indian Wars', 'Napoleonic', 'American Revolutionary War', 'Vietnam War', 'American Civil War', 'Number', 'Trivia', 'Music', 'Korean War', 'City Building', 'Political', 'Math', 'Maze', 'Theme_Food / Cooking', 'Theme_Superheroes', 'Theme_Anime / M

,BGGId,Adventure,Fantasy,Fighting,Environmental,Medical,Economic,Industry / Manufacturing,Transportation,Science Fiction,...,Theme_Fashion,Theme_Geocaching,Theme_Ecology,Theme_Chernobyl,Theme_Photography,Theme_French Foreign Legion,Theme_Cruise ships,Theme_Apache Tribes,Theme_Rivers,Theme_Flags identification
0,1,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,2,0,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,3,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,4,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,5,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0



missing values:


BGGId                             0
Theme_Hanseatic Leagu             0
Theme_Druids                      0
Theme_Music Making & Makers       0
Theme_Psychic Powers              0
Theme_Trash / Garbag              0
Theme_Silk Road                   0
Theme_Cereal Games                0
Theme_Automotive Industry         0
Theme_Video Game Theme: Tetris    0
dtype: int64

In [12]:
themes.drop(columns=["BGGId"]).sum(axis=1).value_counts().head()

1    8287
2    5392
0    4410
3    2566
4     914
Name: count, dtype: int64

In [14]:
inspect_df(subcategories, "subcategories")

--- subcategories ---
shape: (21925, 11)

columns:
['BGGId', 'Exploration', 'Miniatures', 'Territory Building', 'Card Game', 'Educational', 'Puzzle', 'Collectible Components', 'Word Game', 'Print & Play', 'Electronic']

head:


,BGGId,Exploration,Miniatures,Territory Building,Card Game,Educational,Puzzle,Collectible Components,Word Game,Print & Play,Electronic
0,1,0,0,0,0,0,0,0,0,0,0
1,2,0,0,0,1,0,0,0,0,0,0
2,3,0,0,0,0,0,0,0,0,0,0
3,4,0,0,0,0,0,0,0,0,0,0
4,5,0,0,1,0,0,0,0,0,0,0



missing values:


BGGId                     0
Exploration               0
Miniatures                0
Territory Building        0
Card Game                 0
Educational               0
Puzzle                    0
Collectible Components    0
Word Game                 0
Print & Play              0
dtype: int64

In [15]:
subcategories.drop(columns=["BGGId"]).sum(axis=1).value_counts().head()

0    11892
1     8417
2     1461
3      149
4        6
Name: count, dtype: int64

In [16]:
inspect_df(publishers, "publishers")

--- publishers ---
shape: (21925, 1866)

columns:
['Hans im Glück', 'Moskito Spiele', 'Portal Games', 'Spielworxx', 'Stronghold Games', 'Valley Games, Inc.', 'YOKA Games', 'sternenschimmermeer', 'E.S. Lowe', 'Milton Bradley', '999 Games', 'ABACUSSPIELE', 'Descartes Editeur', 'Edge Entertainment', 'Fantasy Flight Games', 'Galakta', 'Hobby Japan', 'Korea Boardgames Co., Ltd.', 'Lacerta', 'Lautapelit.fi', 'Rio Grande Games', 'Smart Ltd', 'Wargames Club Publishing', 'KOSMOS', '3M', 'Avalon Hill Games, Inc.', 'Dujardin', 'Grow Jogos e Brinquedos', 'PS-Games', 'Schmidt France', 'Schmidt International', 'Schmidt Spiele', 'Selecta Spel en Hobby', 'The Avalon Hill Game Co', 'Historien Spiele Galerie (Historien Spielegalerie)', 'Falomir Juegos', 'Family Games, Inc.', 'Gazebo Games UK Ltd.', 'Gigamic', 'Holzinsel', 'Lagoon Games', 'Mattel', 'Vendetta', 'Warfrog Games', 'AMIGO', 'Albi', 'Corfix', 'Hobby World', 'Midgaard Games', 'Brain Games', 'Broadway Toys LTD', 'Brädspel.se', 'Copag Cards', 'Fr

,Hans im Glück,Moskito Spiele,Portal Games,Spielworxx,Stronghold Games,"Valley Games, Inc.",YOKA Games,sternenschimmermeer,E.S. Lowe,Milton Bradley,...,Cacahuete Games,BlackSands Games,Norsker Games,Perro Loko Games,Funko Games,Origame,Deep Print Games,Hidden Industries GmbH,BGGId,Low-Exp Publisher
0,1,1,1,1,1,1,1,1,0,0,...,0,0,0,0,0,0,0,0,1,0
1,0,0,0,0,0,0,0,0,1,1,...,0,0,0,0,0,0,0,0,2,0
2,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,3,1
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,4,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,5,1



missing values:


Hans im Glück            0
Jack Darwid Games        0
Story Factory            0
Magazzini Salani         0
MYBG Co., Ltd.           0
Lex Games                0
Boosterbox.nl            0
Taktyka i Strategia      0
Formosa Force Games      0
Pendragon Game Studio    0
dtype: int64

In [18]:
publisher_counts = publishers.drop(columns=["BGGId", "Low-Exp Publisher"]).sum(axis=1)
publisher_counts.value_counts().sort_index().head(10)

0     2305
1    10579
2     3874
3     1638
4      910
5      613
6      481
7      324
8      260
9      156
Name: count, dtype: int64

In [17]:
inspect_df(ratings_distribution, "ratings_distribution")

--- ratings_distribution ---
shape: (21925, 96)

columns:
['BGGId', '0.0', '0.1', '0.5', '1.0', '1.1', '1.2', '1.3', '1.4', '1.5', '1.6', '1.7', '1.8', '1.9', '2.0', '2.1', '2.2', '2.3', '2.4', '2.5', '2.6', '2.7', '2.8', '2.9', '3.0', '3.1', '3.2', '3.3', '3.4', '3.5', '3.6', '3.7', '3.8', '3.9', '4.0', '4.1', '4.2', '4.3', '4.4', '4.5', '4.6', '4.7', '4.8', '4.9', '5.0', '5.1', '5.2', '5.3', '5.4', '5.5', '5.6', '5.7', '5.8', '5.9', '6.0', '6.1', '6.2', '6.3', '6.4', '6.5', '6.6', '6.7', '6.8', '6.9', '7.0', '7.1', '7.2', '7.3', '7.4', '7.5', '7.6', '7.7', '7.8', '7.9', '8.0', '8.1', '8.2', '8.3', '8.4', '8.5', '8.6', '8.7', '8.8', '8.9', '9.0', '9.1', '9.2', '9.3', '9.4', '9.5', '9.6', '9.7', '9.8', '9.9', '10.0', 'total_ratings']

head:


,BGGId,0.0,0.1,0.5,1.0,1.1,1.2,1.3,1.4,1.5,...,9.2,9.3,9.4,9.5,9.6,9.7,9.8,9.9,10.0,total_ratings
0,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,11.0,5.0,11.0,86.0,3.0,4.0,6.0,8.0,426.0,5352.0
1,2,0.0,0.0,0.0,3.0,0.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,1.0,0.0,0.0,1.0,0.0,17.0,562.0
2,3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,20.0,7.0,4.0,77.0,3.0,1.0,5.0,3.0,477.0,15148.0
3,4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,3.0,342.0
4,5,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,18.0,8.0,4.0,82.0,7.0,9.0,10.0,5.0,905.0,18387.0



missing values:


BGGId    0
0.0      0
7.6      0
7.5      0
7.4      0
7.3      0
7.2      0
7.1      0
7.0      0
6.9      0
dtype: int64

In [11]:
inspect_df(user_ratings, "user_ratings")

--- user_ratings ---
shape: (18942215, 3)

columns:
['BGGId', 'Rating', 'Username']

head:


,BGGId,Rating,Username
0,213788,8.0,Tonydorrf
1,213788,8.0,tachyon14k
2,213788,8.0,Ungotter
3,213788,8.0,brainlocki3
4,213788,8.0,PPMP



missing values:


Username    63
BGGId        0
Rating       0
dtype: int64

### Early observations

**Data structure**

- Each set is connected via `BGGID`.
- `games`, `mechanics`, `themes`, `subcategories`, `publishers` and `rating_distribution` all have 21925 rows, ⭐️ one row per game.
- `user_ratings` is many rows per game.

**Game-level features**

- Games has 47 features describing the game, including:
  - ratings (`AvgRating`, `BayesAvgRating`)
  - difficulty level (`GameWeight`)
  - player counts and playtime, from makers and community
  - popularity metrics (e.g. `NumOwned`, `NumRatings`)
  - categorical flags to describe game type
  - ranking per game type
  - number of alternates, expansions and implementations
- This dataset includes several missing values which will need to be addressed.

**Categorical feature sets (multi-label)**

- `mechanics` (157), `themes` (217), `subcategories` (10) are represented as binary one-hot encoded features.
- 👉 A single game can have more then one mechanic, theme or subcategory. 
- Some games do not have these values recorded.

**Publishers**

- Publisher data is multi-label rather than single-label. 
- Most games have one publisher flagged, but many have multiple, likely reflecting regional editions, reprints, or multiple associated publishers.
- High dimensional (~1800 columns), less directly interpretable

**Ratings data**

- `ratings_distribution`: aggregated counts of ratings (0–10, 0.1 steps)
- `user_ratings`: individual user–game ratings
- Missing usernames present but not critical at this stage


## Decisions for next steps

- Use `BGGId` as the primary join key
- Treat `games` as the core table and base dataset
- Start with a game-level dataset for Project 1: *What makes a highly rated good game?*
  - Join in `mechanics`, `themes`, `subcategories` (multi-label binary feature sets)
  - Use `BayesAvgRating` as the main target variable (more robust that `AvgRating`)
- Keep initial scope focused:
  - Exclude `publishers` for now (high dimensional, less interpretable)
  - Exclude `user_ratings` (reserved for recommendation modelling)
  - Explore simple derived features e.g. number of mechanics, themes, subcategories per game
  - Revisit `user_ratings` in Project 2: *Game recommendation system*
